extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.js

```diff
@@ -558,33 +555,45 @@ export class AmpStoryDevToolsTabPreview extends AMP.BaseElement {
     }).then(() => {
       deviceSpecs.player
         .getElement()
-        .addEventListener('storyNavigation', (event) => {
-          const currTimestamp = Date.now();
-          // Skip sending navigation if it came from a player that's not the latest, and it's within the threshold.
-          // This avoids triggering navigation on programmatic advancements.
-          if (
-            this.lastNavigationData_.player != deviceSpecs.player &&
-            currTimestamp - this.lastNavigationData_.timestamp <
-              NAVIGATION_BETWEEN_PLAYERS_THRESHOLD_MS
-          ) {
-            return;
-          }
-          this.lastNavigationData_ = {
-            player: deviceSpecs.player,
-            timestamp: currTimestamp,
-          };
-          this.devices_.forEach((d) => {
-            if (d != deviceSpecs) {
-              d.player.show(null, event.detail.pageId);
-            }
-          });
-        });
+        .addEventListener('storyNavigation', (event) =>
+          this.onPlayerNavigation_(event, deviceSpecs)
+        );
       deviceSpecs.player.load();
     });
+    this.expectedNavigationEvents_[deviceSpecs.name] = [];
     this.devices_.push(deviceSpecs);
     this.updateDevicesInHash_();
   }
 
+  /**
+   * Triggered when a player emits a storyNavigationEvent.
+   *
+   * A navigation event from a player can come from a user interaction or a previous programmatic call.
+   * Expected navigation events from programmatic calls are stored in `this.expectedNavigationEvents_`,
+   * so they should not be propagated (but deleted from the list of expected events). If an event was not
+   * expected, it means it was user navigation and should be propagated to other players. Context #32189
```

Comment:
You could delete `Context #32189` since that will be visible in commit history and this commenting is robust.

Reply:
True! I added more info on this PR but if this comment is easy to understand, I can remove the context. I do hate complex code that doesn't have proper documentation tho, not the case